pkg/blockchain/rpcLogStreamer.go

```diff
@@ -134,83 +133,194 @@ func (r *RpcLogStreamer) Stop() {
 	r.wg.Wait()
 }
 
-func (r *RpcLogStreamer) watchContract(watcher ContractConfig) {
+func (r *RpcLogStreamer) watchContract(cfg ContractConfig) {
 	var (
-		logger    = r.logger.With(zap.String("contractAddress", watcher.ContractAddress.Hex()))
-		timer     = time.NewTimer(watcher.MaxDisconnectTime)
-		fromBlock = watcher.FromBlock
+		logger    = r.logger.With(zap.String("contractAddress", cfg.Address.Hex()))
+		subLogger = logger.Named("subscription").
+				With(zap.String("contractAddress", cfg.Address.Hex()))
+		isSubEnabled         = false
+		backfillStartBlock   = cfg.FromBlock
+		highestBackfillBlock uint64
 	)
 
-	defer close(watcher.backfillChannel)
-	defer close(watcher.reorgChannel)
-	defer timer.Stop()
+	if backfillStartBlock > 0 {
+		highestBackfillBlock = backfillStartBlock - 1
+	}
+
+	defer close(cfg.eventChannel)
+	defer close(cfg.reorgChannel)
+
+	backfillEndBlock, err := r.client.BlockNumber(r.ctx)
+	if err != nil {
+		logger.Error("failed to get highest block", zap.Error(err))
+		return
+	}
 
-	logger.Info("Starting watcher", zap.Uint64("fromBlock", fromBlock))
+	innerBackfillCh, cancelBackfill, err := r.buildBackfillChannel(cfg, backfillStartBlock)
+	if err != nil {
+		logger.Error("failed to retrieve historical logs", zap.Error(err))
+		return
+	}
+
+	innerSubCh, err := r.buildSubscriptionChannel(cfg, subLogger)
+	if err != nil {
+		logger.Error("failed to subscribe to contract", zap.Error(err))
+		return
+	}
+
+	logger.Info("Starting watcher")
 
 	for {
 		select {
 		case <-r.ctx.Done():
-			logger.Debug("Stopping watcher")
+			logger.Debug("Context cancelled, stopping watcher")
 			return
-		case <-timer.C:
-			logger.Fatal(
-				"Max disconnect time exceeded. Node might drift too far away from expected state. Shutting down...",
-			)
 
-		case reorgBlock, open := <-watcher.reorgChannel:
+		case reorgBlock, open := <-cfg.reorgChannel:
 			if !open {
 				logger.Debug("Reorg channel closed")
 				return
 			}
-			fromBlock = reorgBlock
+			backfillStartBlock = reorgBlock
 			logger.Warn(
 				"Blockchain reorg detected, resuming from block",
-				zap.Uint64("fromBlock", fromBlock),
+				zap.Uint64("fromBlock", backfillStartBlock),
 			)
-			timer.Reset(watcher.MaxDisconnectTime)
-
-		default:
-			logs, nextBlock, err := r.GetNextPage(watcher, fromBlock)
-			if err != nil {
-				logger.Error(
-					"Error getting next page",
-					zap.Uint64("fromBlock", fromBlock),
-					zap.Error(err),
-				)
-				time.Sleep(ERROR_SLEEP_TIME)
+
+		case log, open := <-innerSubCh:
+			if !open {
+				subLogger.Fatal("subscription channel closed, closing watcher")
```

Comment:
does fatal call panic?

This also has to close all channels. Or the better idea might be to call defer on the cancel functions after creation.

Reply:
Good catch as well, switching to logger.Error + return.